### PR TITLE
Adding a `pscheduler metrics` command for stats

### PR DESCRIPTION
**Pull request by tonin**
Outputs all the statistics available through the API.  It has 2 output format: json (default) and
prometheus to be used with prometheus-node-exporter textfile exporter.

### CODE REVIEW

**Comment by tonin** [comment]
Output of this new `metrics` command is for JSON:
```
$ pscheduler metrics | jq
{
  "http-queue": {
    "length": 0,
    "backlog": 0
  },
  "paused": {
    "is_paused": false
  },
  "runs": {
    "missed": 0,
    "running": 2406,
    "cleanup": 36,
    "preempted": 0,
    "overdue": 0,
    "failed": 201,
    "pending": 13,
    "nonstart": 3262,
    "finished": 56207,
    "on-deck": 2
  },
  "archiving": {
    "upcoming": 0,
    "backlog": 1
  }
}
```

and for Prometheus:
```
$ pscheduler metrics --format prometheus
# HELP perfsonar_pscheduler_archiving pScheduler archiving queue
# TYPE perfsonar_pscheduler_archiving gauge
perfsonar_pscheduler_archiving{state="upcoming"} 0
perfsonar_pscheduler_archiving{state="backlog"} 1
# HELP perfsonar_pscheduler_paused pScheduler is paused
# TYPE perfsonar_pscheduler_paused gauge
perfsonar_pscheduler_paused{state="is_paused"} 0
# HELP perfsonar_pscheduler_runs Number of pScheduler runs in the different states
# TYPE perfsonar_pscheduler_runs gauge
perfsonar_pscheduler_runs{state="failed"} 201
perfsonar_pscheduler_runs{state="cleanup"} 36
perfsonar_pscheduler_runs{state="nonstart"} 3262
perfsonar_pscheduler_runs{state="finished"} 56207
perfsonar_pscheduler_runs{state="overdue"} 0
perfsonar_pscheduler_runs{state="pending"} 13
perfsonar_pscheduler_runs{state="on-deck"} 2
perfsonar_pscheduler_runs{state="preempted"} 0
perfsonar_pscheduler_runs{state="running"} 2406
perfsonar_pscheduler_runs{state="missed"} 0
# HELP perfsonar_pscheduler_http_queue pScheduler HTTP queue
# TYPE perfsonar_pscheduler_http_queue gauge
perfsonar_pscheduler_http_queue{state="backlog"} 0
perfsonar_pscheduler_http_queue{state="length"} 0
```

File: pscheduler-core/pscheduler-core/metrics [hunk 1]
```diff
@@ -0,0 +1,132 @@
+#!/usr/bin/python3
```
Comment (mfeit-internet2) [inline pscheduler-core/pscheduler-core/metrics:1]: Program needs to be added to the `COMMANDS` list in the `Makefile`.

File: pscheduler-core/pscheduler-core/metrics [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                      help="The output format to use, either json (default) or prometheus",
+                      default="json",
+                      action="store", type="string",
+                      dest="format")
+opt_parser.add_option("--timeout", "-W",
+                      help="How long to wait for the server to respond, in seconds (default 5)",
+                      default=5,
+                      action="store", type="int",
+                      dest="timeout")
+
```
Comment (mfeit-internet2) [inline pscheduler-core/pscheduler-core/metrics:81]: I recommend adding a `--host` option (see other programs, like `plot-schedule`) and using it in forming URLs for the API.  See comment below.

File: pscheduler-core/pscheduler-core/metrics [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    pscheduler.fail("Format must be either json or prometheus")
+
+log = pscheduler.Log(verbose=False, debug=False, quiet=True, propagate=True)
+
+# Collect statistics
+all_stats = {}
+for k,v in pscheduler_stats.items():
+    my_stat = {}
+    for state in v["state"]:
+        url = pscheduler.api_url("localhost", v["url"] + state)
```
Comment (mfeit-internet2) [inline pscheduler-core/pscheduler-core/metrics:98]: Use the value of the `host` option here.  If the default for that is `None`, `api_url()` will construct a URL for the local host if the option is not specified.    (FYI, There's an `api_local_host` call in the library.)

File: pscheduler-core/pscheduler-core/metrics [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                my_stat[state] = result
+            else:
+                my_stat = result
+
+    all_stats[k] = my_stat
+
+# Output the collected statistics
+output = ""
+if options.format == "json":
+    output = json.dumps(all_stats)
```
Comment (mfeit-internet2) [inline pscheduler-core/pscheduler-core/metrics:115]: If you want to dump JSON, declare success and exit, call `pscheduler.success_json(all_stats)` for a one-stop solution.

There's also a `json_dump` in the library.

File: pscheduler-core/pscheduler-core/metrics [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        output += '# HELP {} {}\n'.format(pscheduler_stats[k]["prometheus-name"], pscheduler_stats[k]["desc"])
+        output += '# TYPE {} {}\n'.format(pscheduler_stats[k]["prometheus-name"], pscheduler_stats[k]["type"])
+        for state, value in v.items():
+            if pscheduler_stats[k]["state"] != "":
+                output += '{}{{state="{}"}} {}\n'.format(pscheduler_stats[k]["prometheus-name"], state, int(value))
+            else:
+                output += '{} {}\n'.format(pscheduler_stats[k]["prometheus-name"], int(value))
+
+else:
+    output = "Unknown output format"
```
Comment (mfeit-internet2) [inline pscheduler-core/pscheduler-core/metrics:128]: Call `pscheduler.fail("Error Message")` here.  The error will go to stderr and the program will exit 1.

File: pscheduler-core/pscheduler-core/metrics [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+
+print(output)
```
Comment (mfeit-internet2) [inline pscheduler-core/pscheduler-core/metrics:131]: Call `pscheduler.succeed(output)` here to exit cleanly.

**Review by mfeit-internet2 — CHANGES_REQUESTED** [review]
Just a bunch of small stuff in the comments.  Otherwise appears to work as advertised.

**Comment by tonin** [comment]
Thanks for your review Mark.  I made the changes, let me know if you want anything else.

**Review by mfeit-internet2 — APPROVED** [review]
Looks good.  I'll pull it and delete the branch.

**Pull request merged**